Login

MartyOfMCA opened the pull request:
Quite a lot has been done in this PR so to make things a lot easier for you, kindly review the changes against these:

* Login feature completed so users can now be authenticated
* Unit test class added to test all the functionalities in the Login class (well almost all the functionalities)
* Resolved the problem in the AppDbContext.cs file
* Condensed the migrations into a single initial migration

**Disclaimer**
Because we aren't using the ticket numbers it was challenging for me to indicate which portions of Patrick's code my code broke when I was committing my changes. But I made sure I resolved all those scenarios as well (I think there was only one instance I had to do that - line 160 of Register.cs). I know I should have informed Patrick about this but I had to just add some 5 characters to his code for my project to function as intended (it's not like they're tightly coupled or anything like that).

Whenever you pull this code, right after you branch off, make sure to head over to the Program.cs and change new Login() to new Main() or new Register() or whichever you deem fit.

Commit:
feat: add login interface

add form controls to accept user input for both Employee ID and
passwords. User input is validated as required before the user
is given permission to the system

Commit:
refactor: add Enums.cs for all flags in codebase

add the Enums.cs in the FPIS.Utils namespace to store all flags for the
various actions we might perform. This is extremely helpful in situations
where we might want to perform different actions based on the feedback
obtained from a function call. For example, during authentication we
might have to display a message when no employee id is provided. Or
when a password is not provided. Or when the authentication fails or
passes.

Commit:
fix: correct the namespace for Utils.cs

fix the namespace for the Utils source code file because it was preventing
the code from compiling. It was ealier in the FPIS namespace.

Commit:
fix: correct the namespace for function call

update the namespace to the Utils.cs class when invoking the HashPassword
function during the registration process. This had to be done because a new
source code file had been added to the Utils namespace.

Commit:
docs: add documentation to Enums.cs

add a little description to this enumeration. Descriptions were
provided for the individual constants but I forgot to provide
one for the container itself lol

Commit:
refactor: cleanup Login.cs with right functions

made code changes in the AuthenticateUser and UserProvidedData functions
to suit the changes made. These functions are more modular and perform
better than earlier because now they only have to do just one thing and
return some token or value to the caller to decide what to do.

Commit:
test: add unit test for Login.cs

add the LoginTest.cs unit test file with two test cases to easily determine
the functionality of the codes in the Login.cs source code file. Also
added public functions to easily test the functions in the Login.cs file. These
public functions would have to be deleted later on because they are only meant
for us and not for production.

Commit:
fix: remove migrations

remove all the migrations created earlier to condense them into a single initial
migration. Also resolves the problem with AppDbContext.cs by removing code in
line 7.

Commit:
chore: add initial migration

add the very first migration that comprises all the models used in the
system as we agreed from our initial discussions on the database design.

denkyipatrick commented on FPIS/Program.cs:
```diff
@@ -13,7 +13,7 @@ static void Main()
             // To customize application configuration such as set high DPI settings or default font,
             // see https://aka.ms/applicationconfiguration.
             ApplicationConfiguration.Initialize();
-            Application.Run(new Main());
```
Can you revert this back to new Main()?

MartyOfMCA replied:
Alright... will do

denkyipatrick commented on FPIS/Utils/Enums.cs:
```diff
@@ -0,0 +1,27 @@
+        /// <summary>
+        /// User didn't provide a username
+        /// </summary>
+        NOUSERNAME,
```
Can you use a consistent naming convention? For readability 😄  
Suggestion: `NOUSERNAME` => `NO_USERNAME`

MartyOfMCA replied:
Lol... my bad. I'll use the one with underscores.

MartyOfMCA replied:
I meant the one WITHOUT

denkyipatrick commented on FPIS/Utils/Enums.cs:
```diff
@@ -0,0 +1,27 @@
+        /// <summary>
+        /// User didn't provide any password
+        /// </summary>
+        NOPASSWORD,
```
`NOPASSWORD` => `NO_PASSWORD` just like above.

denkyipatrick commented on FPIS/Views/Login.cs:
```diff
@@ -11,6 +13,69 @@ public Login()
+        }
+
+        //TODO: Delete this test code
+        public LoginAuth TestAuthenticateUser(string employeeId, string password) => AuthenticateUser(employeeId, password);
```
If this is not needed by others, you can remove it 😄

MartyOfMCA replied:
You do realize the unit tests will fail should there be the need to run them. This code is probably not useful because I already run the tests and they work just fine. But perhaps one of you might want to run the test to see how it works. But I'll do away with them anyways

denkyipatrick commented on FPIS/Views/Login.cs:
```diff
@@ -11,6 +13,69 @@ public Login()
+        {
+            // Seperately check the username and passwords to show different error messages
+            bool dataProvided = UserProvidedData(employeeId);
+            if (!(dataProvided))
```
Please put `{}` after the `if (condition)` to avoid code smells. This usually causes difficult-to-debug errors.

MartyOfMCA replied:
Alright

denkyipatrick commented on FPIS/Views/Login.cs:
```diff
@@ -11,6 +13,69 @@ public Login()
+        }
+
+        //TODO: Delete this test code
+        public bool TestUserProvidedData(string userInput) => UserProvidedData(userInput);
```
Delete this code 😄

Commit:
fix: implement change request

implement the code changes requested by Patrick. Changed the startup screen
back to Main. Made sure identifiers stick with a single naming convention.
Took away the test codes and added blocks for the if statements I wrote in
the Login.cs.

Commit:
fix: implement change request

implement the code changes requested by Patrick. Very similar to previous commit.
It turns out I forgot to save the recent changes in the editor so I winded up
commiting not up-to-date code. Made sure identifiers stick with a single naming convention.
Added blocks for the if statements I wrote in the Login.cs.

denkyipatrick commented on FPIS/Views/Login.cs:
```diff
@@ -11,6 +13,69 @@ public Login()
+            bool dataProvided = UserProvidedData(employeeId);
+            if (!dataProvided)
+            {
+                return LoginAuth.NOUSERNAME;
```
There is no variable called `NOUSERNAME` as you have changed it to `NO_USERNAME` and the others

MartyOfMCA replied:
Hmm... On it

denkyipatrick commented on FPIS/Models/AppDbContext.cs:
```diff
@@ -4,11 +4,9 @@ namespace FPIS.Models
         protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
         {
-            optionsBuilder.UseNpgsql("Host=localhost;Port=5432;Database=FPIS;Username=postgres;Password=eugene");
+            optionsBuilder.UseNpgsql("Host=localhost;Port=5432;Database=FPIS;Username=postgres;Password=pa$$w0rd");
```
Don't push your credentials.

Commit:
fix: implement change request

implement the code changes requested by Patrick. Finally resolved all the
issues preventing the code from compiling

Commit:
fix: implement change request

implement all the necessary code changes requested

denkyipatrick commented:
Code Looks good. 
But you should know that the `Registration Feature` is no longer working because of the `DesignationId` referential integrity.
I have merged the pull request anyway because we need to continue working.
Another task should be created to fix the `DesignationId` issue with regards to `registration`.
